File: src/kernel/boot/log.zig
//! Boot Sequence Logging.
//!
//! Centralized boot progress messages with consistent formatting.
//!
//! Format: "[N/T] name message"
//! - N/T: stage progress
//! - name: subsystem, padded to a consistent column
//! - message: description

const builtin = @import("builtin");
const console = @import("../console/console.zig");
const hwinfo = @import("../hwinfo/hwinfo.zig");
const pmm_mod = @import("../pmm/pmm.zig");

/// Total boot stages. Update when adding/removing stages.
pub const STAGES: u8 = 10;

const MB: usize = 1024 * 1024;

const STAGE_NAMES = [_][]const u8{
    "uart",
    "trap",
    "mmu",
    "virt",
    "dtb",
    "pmm",
    "trace",
    "clock",
    "task",
    "idle",
};

const STAGE_NAME_WIDTH: usize = blk: {
    var max: usize = 0;
    for (STAGE_NAMES) |name| {
        if (name.len > max) max = name.len;
    }
    break :blk max;
};

pub fn header() void {
    console.print("Bullfinch\n\n");
}

/// Print stage prefix: "[N/T] name " with padded name column.
/// Use for stages with dynamic content. Caller must print message and newline.
fn stagePrefix(n: u8, name: []const u8) void {
    console.print("[");
    if (n < 10) console.print("0");
    console.printDec(n);
    console.print("/");
    if (STAGES < 10) console.print("0");
    console.printDec(STAGES);
    console.print("] ");

    printPadded(name, STAGE_NAME_WIDTH);
    console.print(" ");
}

/// Print a boot stage message: "[N/T] name message\n"
fn stage(n: u8, name: []const u8, message: []const u8) void {
    stagePrefix(n, name);
    console.print(message);
    console.print("\n");
}

pub fn uart() void {
    stage(1, "uart", "console ready");
}

pub fn trap() void {
    stage(2, "trap", "exception handlers set");
}

pub fn mmu() void {
    stage(3, "mmu", "virtual memory enabled");
}

pub fn virt() void {
    stage(4, "virt", "low addresses unmapped");
}

pub fn dtb() void {
    const hw = &hwinfo.info;
    stagePrefix(5, "dtb");
    console.printDec(hw.cpu_count);
    console.print(" cpus, ");
    console.printDec(hw.total_memory / MB);
    console.print(" MB\n");
}

pub fn pmm() void {
    const arenas = pmm_mod.arenaCount();
    const pages = pmm_mod.totalPages();
    const pages_k = pages / 1000;

    stagePrefix(6, "pmm");
    console.printDec(arenas);
    console.print(if (arenas == 1) " arena, " else " arenas, ");
    console.printDec(pages_k);
    console.print("k pages\n");
}

pub fn trace() void {
    stage(7, "trace", "ring ready");
}

pub fn clock() void {
    stage(8, "clock", "timer ready");
}

pub fn task() void {
    stage(9, "task", "scheduler ready");
}

pub fn idle() void {
    stage(10, "idle", "entering idle thread");
}

fn printPadded(text: []const u8, width: usize) void {
    console.print(text);
    var pad: usize = width -| text.len;
    while (pad > 0) : (pad -= 1) console.print(" ");
}

File: kernel/src/startup/log.rs
//! Boot sequence logging.
//!
//! Keep early boot output compact and stable for smoke-log comparison.

use kernel::hwinfo::HardwareInfo;

use crate::console::Console;

pub const STAGES: usize = 10;

const MEGABYTE: usize = 1024 * 1024;
const STAGE_NAME_WIDTH: usize = 9;

pub fn header() {
    Console::new().print("[BOOT:STARTED]\n");
}

pub fn uart() {
    stage(1, "boot", "console ready");
}

pub fn trap() {
    stage(2, "trap", "exception handlers set");
}

pub fn mmu() {
    stage(3, "mmu", "virtual memory enabled");
}

pub fn virt() {
    stage(4, "mmu", "low addresses unmapped");
}

pub fn dtb(info: &HardwareInfo) {
    let mut out = Console::new();
    stage_prefix(&mut out, 5, "hwinfo");
    out.print_dec_usize(info.cpu_count);
    out.print(" cpus, ");
    out.print_dec_usize(info.total_memory / MEGABYTE);
    out.print(" MB\n");
    if info.dropped_memory_regions > 0 {
        stage_prefix(&mut out, 5, "hwinfo");
        out.print("warning: dropped ");
        out.print_dec_usize(info.dropped_memory_regions);
        out.print(" memory region(s) (limit reached)\n");
    }
    if info.dropped_reserved_regions > 0 {
        stage_prefix(&mut out, 5, "hwinfo");
        out.print("warning: dropped ");
        out.print_dec_usize(info.dropped_reserved_regions);
        out.print(" reserved region(s) (limit reached)\n");
    }
}

pub fn pmm() {
    let mut out = Console::new();
    stage_prefix(&mut out, 6, "pmm");
    out.print_dec_usize(kernel::pmm::arena_count());
    out.print(" arenas, ");
    out.print_dec_usize(kernel::pmm::total_pages());
    out.print(" pages\n");
}

pub fn trace() {
    stage(7, "trace", "ring ready");
}

pub fn clock() {
    stage(8, "clock", "timer ready");
}

pub fn task() {
    stage(9, "task", "scheduler ready");
}

pub fn idle() {
    stage(10, "task", "entering idle thread");
}

fn stage(number: usize, name: &str, message: &str) {
    let mut out = Console::new();
    stage_prefix(&mut out, number, name);
    out.print(message);
    out.print("\n");
}

fn stage_prefix(out: &mut Console, number: usize, name: &str) {
    out.print("[");
    print_two_digits(out, number);
    out.print("/");
    print_two_digits(out, STAGES);
    out.print("] ");
    print_padded(out, name, STAGE_NAME_WIDTH);
    out.print(" ");
}

fn print_two_digits(out: &mut Console, value: usize) {
    if value < 10 {
        out.print("0");
    }
    out.print_dec_usize(value);
}

fn print_padded(out: &mut Console, text: &str, width: usize) {
    out.print(text);
    for _ in text.len()..width {
        out.print(" ");
    }
}
